Thea: Fisherman and Batch Signing

Gauthamastro opened the pull request:
## Describe your changes

This PR raises the final changes to core Thea logic by adding Fisherman model for deposits and batch signing for withdrawals

Gauthamastro committed:
Add on-chain signature aggregation

Gauthamastro committed:
Added Fisherman and dealing with misbehaviour logic

Gauthamastro committed:
Thea core logic completed

Gauthamastro commented:
@zktony Please add tests for this PR. Let me know if you need any help

Gauthamastro committed:
Fix runtime

zktony committed:
Added chnages for eth

Gauthamastro committed:
Update thea logic

Gauthamastro committed:
Added changes for Ethereum  (#896)

## Describe your changes
Changed types and added new extrinsic to support Eth.

Gauthamastro committed:
cargo fmt

Gauthamastro committed:
fix thea tests

Gauthamastro committed:
fix thea-executor tests

Gauthamastro committed:
Update Thea message handler

Gauthamastro committed:
Fix tests

Gauthamastro committed:
Fix benchmarks

Gauthamastro committed:
Resolve clippy errors

Gauthamastro committed:
cargo fmt

Gauthamastro committed:
Fix benchmarking

zktony committed:
Added sign_prehash

zktony committed:
Removed comment

Gauthamastro committed:
Changed sign to sign_prehash (#897)

## Describe your changes
Changed sign to sign_prehash.

Gauthamastro committed:
fix cargo test

serhii-temchenko commented on pallets/thea-executor/src/lib.rs:
```diff
@@ -402,6 +403,30 @@ pub mod pallet {
+
+		#[pallet::call_index(6)]
+		#[pallet::weight(<T as Config>::WeightInfo::parachain_withdraw(1))] // TODO: @zktony Benchmarking
+		pub fn ethereum_withdraw(
```
No tests for this extrinsic

zktony replied:
Added one test case to test the happy path. Parachain_withdraw tests already cover other cases.

serhii-temchenko commented on pallets/thea-executor/src/lib.rs:
```diff
@@ -402,6 +403,30 @@ pub mod pallet {
 			Ok(())
 		}
+
+		#[pallet::call_index(6)]
+		#[pallet::weight(<T as Config>::WeightInfo::parachain_withdraw(1))] // TODO: @zktony Benchmarking
+		pub fn ethereum_withdraw(
+			origin: OriginFor<T>,
+			asset_id: u128,
+			amount: u128,
+			beneficiary: H160,
+			pay_for_remaining: bool,
+			pay_with_tokens: bool,
+		) -> DispatchResult {
```
Would be nice to have docblock here

zktony replied:
done

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -86,31 +87,57 @@ pub trait TheaWeightInfo {
 }
 
-#[frame_support::pallet]
+#[frame_support::pallet(dev_mode)]
```
Should this `dev_mode` be removed?

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -205,31 +336,75 @@ pub mod pallet {
+			let config = <NetworkConfig<T>>::get(payload.network);
+
+			if stake < config.min_stake {
+				return Err(Error::<T>::NotEnoughStake.into())
```
Should be tested this condition

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -296,6 +481,140 @@ pub mod pallet {
+		/// Signed outgoing messages
+		#[pallet::call_index(6)]
+		#[pallet::weight(10_000)] // TODO: @zktony Benchmarking.
+		pub fn submit_signed_outgoing_messages(
```
It should be tested

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -296,6 +481,140 @@ pub mod pallet {
+		#[pallet::call_index(7)]
+		#[pallet::weight(10_000)] // TODO: @zktony Benchmarking
+		#[transactional]
+		pub fn report_misbehaviour(
```
Should be tested

serhii-temchenko commented on pallets/thea/src/lib.rs:
```diff
@@ -296,6 +481,140 @@ pub mod pallet {
+		#[pallet::call_index(8)]
+		#[pallet::weight(10_000)] // TODO: @zktony Benchmarking
+		#[transactional]
+		pub fn handle_misbehaviour(
```
Should be tested

serhii-temchenko commented on primitives/thea/src/types.rs:
Please add documentation for each new member (pub api) of this file

zktony committed:
Added benchmarks

zktony committed:
Added on_init benchmark

zktony committed:
Added benchmark code for Thea Pallet and thea Executor Pallet (#905)

## Describe your changes
Added benchmark code for Thea Pallet and thea Executor Pallet.

serhii-temchenko committed:
feat(weights): Updated thea weights

zktony committed:
Resolved all the comments

zktony committed:
Resolved all the comments (#906)

## Describe your changes
Resolved all the comments.

zktony committed:
Fixed ci issues

zktony committed:
Fix CI Issues

serhii-temchenko committed:
feat(weights): Updated weights for thea related pallets

zktony committed:
Fixed tests

zktony committed:
Merge branch 'gj/thea-final' of https://github.com/Polkadex-Substrate/Polkadex into gj/thea-final

serhii-temchenko commented on .github/workflows/ci.yml:
```diff
@@ -115,7 +115,7 @@ jobs:
         run: cargo clippy -- -D warnings
       - name: Test Project
-        run: RUSTFLAGS="-D warnings" cargo test --workspace
+        run: cargo test --workspace
```
Can this be moved back?

serhii-temchenko commented on pallets/thea-executor/src/mock.rs:
```diff
@@ -129,6 +129,8 @@ impl pallet_assets::Config for Test {
 	type Extra = ();
 	type CallbackHandle = ();
 	type WeightInfo = ();
+	// #[cfg(feature = "runtime-benchmarks")]
+	// type BenchmarkHelper = AssetU128;
```
Please remove

serhii-temchenko commented on pallets/thea-message-handler/src/lib.rs:
```diff
@@ -181,44 +187,50 @@ pub mod pallet {
 		/// Handles the verified incoming message
 		#[pallet::call_index(1)]
-		#[pallet::weight(<T as Config>::WeightInfo::incoming_message())]
+		#[pallet::weight(<T as Config>::WeightInfo::incoming_message())] // TODO: @zktony Benchmark again for updated logic
```
I assume this was updated? Can we remove todo?

Gauthamastro commented:
@zktony , The pending tests.

- [x] on_init in thea
- [x] paramaterization of on_initialize function in thea 
- [x] validate_signed_outgoing_message
- [x] submit_incoming_message
- [x] match statement inside incoming_message extrinsic in thea message handler.
- [x] validate_incoming_message in thea message handler ( else blocks are not tested )
- [x] on_initialize  in thea executor
- [x] paramaterization of on_initialize function in thea executor
- [x] add_signature() else condition, contains_signature() methods of SignedMessage in types
- [x] amount_in_foreign_decimals() method of Deposit type in types.rs

zktony committed:
Fixed tests

zktony committed:
Fixed CI

Gauthamastro committed:
Update the correct runtime spec_version

Gauthamastro committed:
Update the correct parachain runtime spec_version

serhii-temchenko committed:
feat(weights): Updated thea related pallets weights
